docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,192 @@
+_Aug 2022_
+
+## Background
+
+Python 3.0 introduced optional function annotations ([PEP 3107](https://peps.python.org/pep-3107/)), which were later codified for use in static type checking around the release of Python 3.5 ([PEP 484](https://peps.python.org/pep-0484/)).
+To some degree, type annotations and static type checking have become an integral part of many Python development workflows, and to this end we have added annotations in a number of places throughout the JAX API.
+The current state of type annotations in JAX is a bit patchwork, and efforts to add more have been hampered by more fundamental design questions.
+This doc attempts to summarize those issues and generate a roadmap for the goals and non-goals of type annotations in JAX.
+
+Why do we need such a roadmap? Better/more comprehensive type annotations are a frequent request from users, both internally and externally.
+In addition, we frequently receive pull requests from external users (for example, [PR #10322](https://github.com/google/jax/pull/10322), [PR #9917](https://github.com/google/jax/pull/9917)) seeking to improve JAX's type annotations: it's not always clear to the JAX team member reviewing the code whether such contributions are beneficial, particularly when they introduce complex Protocols to address the challenges inherent to full-fledged annotation of JAX's use of Python.
+This document details JAX's goals and recommendations for type annotations within the package.
+
+## Why type annotations?
+
+There are a number of reasons that a Python project might wish to annotate their code-base; we'll summarize them in this document as Level 1, Level 2, and Level 3.
+
+### Level 1: Annotations as documentation
+
+When originally introduced in [PEP 3107](https://peps.python.org/pep-3107/), type annotations were motivated partly by the ability to use them as concise, inline documentation of function parameter types and return types. JAX has long utilized annotations in this manner; an example is the common pattern of creating type names aliased to `Any`. An example can be found in `lax/slicing.py` [[source](https://github.com/google/jax/blob/2bc3e39cd9104071ee39dacac22abd51b94eb27e/jax/_src/lax/slicing.py#L47-L58)]:
+
+```python
+Array = Any
+Shape = core.Shape
+
+def slice(operand: Array, start_indices: Sequence[int],
+          limit_indices: Sequence[int],
+          strides: Optional[Sequence[int]] = None) -> Array:
+  ...
+```
+
+For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all), but it does serve as a form of useful in-code documentation for the developer.
+
+For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type `Any`), so the documentation benefit does not go beyond the source code.
```

Comment:
FYI the autodoc extension has [some control](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_type_aliases) over this.

Another trick we [use in Haiku](https://github.com/deepmind/dm-haiku/blob/78c2984969d5b360c19a12bb7eb609712b8312f8/docs/conf.py#L36) docs is adding `from __future__ import annotations` to each .py source when building docs. Not pretty, but gets the job done :)

Reply:
Thanks! I added this to the discussion.